src/webnn/native/dml/DMLUtils.h

```diff
@@ -144,32 +200,86 @@ namespace webnn::native::dml {
                 D3D12_TEXTURE_LAYOUT_ROW_MAJOR,  flags};
     };
 
-    template <typename T>
-    T RoundUpToMultiple(T value, T multiple) {
-        static_assert(std::is_integral_v<T>);
-
-        T remainder = value % multiple;
-        if (remainder != 0) {
-            value += multiple - remainder;
-        }
+    struct CompiledGraph {
+        CompiledGraph(const ComPtr<IDMLDevice>& device,
+                      const DML_GRAPH_DESC& graphDesc,
+                      const std::vector<std::shared_ptr<InputNode>>& inputs,
+                      const std::vector<Node>& outputs,
+                      DML_EXECUTION_FLAGS flag = DML_EXECUTION_FLAG_NONE)
+            : inputNodes(inputs), outputNodes(outputs) {
+            ComPtr<IDMLDevice1> device1;
+            WEBNN_CHECK(device.Get()->QueryInterface(IID_PPV_ARGS(&device1)));
+            WEBNN_CHECK(device1->CompileGraph(&graphDesc, flag, IID_PPV_ARGS(&compiledOperator)));
+            IDMLCompiledOperator* compiledOperators[] = {compiledOperator.Get()};
+            WEBNN_CHECK(
+                device->CreateOperatorInitializer(ARRAYSIZE(compiledOperators), compiledOperators,
+                                                  IID_PPV_ARGS(&compiledOperatorInitializer)));
+        };
 
-        return value;
-    }
+        // IDMLCompiledOperator represents the DirectML graph's output which need to be initialized
+        // by IDMLOperatorInitializer.
+        ComPtr<IDMLCompiledOperator> compiledOperator;
+        ComPtr<IDMLOperatorInitializer> compiledOperatorInitializer;
+        std::vector<std::shared_ptr<InputNode>> inputNodes;
+        std::vector<Node> outputNodes;
+    };
 
     // An adapter called the "Microsoft Basic Render Driver" is always present. This adapter is a
     // render-only device that has no display outputs.
-    HRESULT IsWarpAdapter(IDXGIAdapter1* pAdapter, bool* isWarpAdapter);
-
-    void InitD3D12(ComPtr<ID3D12GraphicsCommandList>& commandList,
-                   ComPtr<ID3D12CommandQueue>& commandQueue,
-                   ComPtr<ID3D12CommandAllocator>& commandAllocator,
-                   ComPtr<ID3D12Device>& D3D12Device,
-                   DXGI_GPU_PREFERENCE gpuPreference,
-                   bool useGpu);
-
-    void CloseExecuteResetWait(ComPtr<ID3D12GraphicsCommandList> commandList,
-                               ComPtr<ID3D12CommandQueue> commandQueue,
-                               ComPtr<ID3D12CommandAllocator> commandAllocator,
-                               ComPtr<ID3D12Device> D3D12Device);
+    bool IsWarpAdapter(IDXGIAdapter1* pAdapter);
+
+    uint64_t RoundUpToMultiple(uint64_t value, uint64_t multiple);
+
+    class Device {
```

Comment:
nit: consider moving this out into `DeviceDML.cpp`, as using one large header for the whole DML backend will be difficult to maintain.